Timeline: Vertical Detail and Timeline View

NghiaTranUIT opened the pull request:
### 📒 Description
This PR will introduce the layout and basic logic for Timeline CollectionView

### 🕶️ Types of changes
**New feature** (non-breaking change which adds functionality)

### 🤯 List of changes
- [x] Introduce Timeline CollectionView and all relevant stuff (Views and Model)
- [x] Introduce TimelineFlowLayout: Able to render Time Label, Time Entry, Divider and Activity 
- [x] Logic for rendering Time Entry correctly. Support `overlap`, if the time entry is overlap, it will move to next columns
- [x] Fetch timeline data: Modify the timeline in gui.cc to get all time entry in this day rather than Chunks (15 mins)
- [x] Date Picker logic
- [x] Zoom logic (-, +)
- [x] For some reason, TimeEntry in Timeline in Library doesn't contain any Project Info
- [x] Hover Time Entry View
- [x] Empty Space for creating new time Entry
- [x] Allow solving overlaps. Clicking on overlap shows a context menu with 3 choices: "change first entry stop time", "change last entry start time" and "create a new entry of overlap" => Not logic yet ❌ 
- [x] Hovering items will show the info about the item (https://zpl.io/VQv3jz4)
- [x] Clicking on items will show edit popup for the item

### 👫 Relationships
Close #2892 
Close #3017

### 🔎 Review hints
### Drawing 
1. Create bunch of TEs with various durations 
2. Check if the Timeline draws the TE properly or not
=> If the TE is overlapped -> The Timeline should draw in next columns 
3. Zoom on, zoom it 
=> TEs shouldn't to closed => It should has a minimum gaps between each other
4. If the TE's duration is less than 1 min
=> IT should draw at 2 pixel
5. If the TE's duration is 2 mins (for instance), but we zoom out 
=> It should draw at 2 pixel
6. The color of TE
=> This color should match with the color of its project

### Insert the Empty Time Entry
1. In the first columns of Timeline, if there is a gap
=> Make sure the Timeline inserts the "Empty Time Entry" properly 
2. If zoom out, and the gap is too small
=> The "Empty Time Entry" should be hidden

### Change day
1. If we changes date
=> The Timeline should re-draw, and has correct number of TE in this days (in comparing with the Time Entry List)

### Hover and Editor on TE
1. If the mouse hover on any TE in Timeline
=> IT should show the hover views (check design)
2. If we click on any TE
=> It should show the Editor on it 
3. If the Editor is opening, move the mouse around
=> The editor shouldn't close 

### Update TE's info
1. If we update the TE's info in Editor
=> IT should update immediately (For instance, change the project with different color => The TE color should be updated)
2. Change date
=> Should render again

NghiaTranUIT commented:
@IndrekV Can you elaborate what it happen in those Context menu: ```Allow solving overlaps. Clicking on overlap shows a context menu with 3 choices: "change first entry stop time", "change last entry start time" and "create a new entry of overlap"```

IndrekV commented:
- "change first entry stop time"
```
From 2 entries that overlap the entry with a smaller start time is updated so that the stop time of that entry is the same as the start time if the later entry
```
- "change last entry start time"
```
From 2 entries that overlap the entry with a later start time is updated so that the start time of that entry is the same as the stop time if the entry that was started before
```

```
- "create a new entry of overlap"
I'm not sure this option is needed. I think we can skip this.

NghiaTranUIT commented:
@IndrekV this PR is really for reviewing in the first phrase. Let check it out 👍 

## Known-issues
- [ ] After creating new time entry, the Timeline view will be reloaded from the call in library -> Cause the editor jumps 
- [ ] Logic behind the "Resolve overlapped Time Entry"

NghiaTranUIT commented:
Updated Review note 👍

NghiaTranUIT commented on src/gui.cc:
```diff
@@ -494,6 +494,21 @@ void GUI::DisplayTimeline(
+    TogglTimeEntryView *first_entry = nullptr;
+    time_t start_day = datetime.timestamp().epochTime();
+    time_t end_day = start_day + 86400; // one day
+    for (unsigned int i = 0; i < entries_list.size(); i++) {
```
Get all TEs in this selected day from library, then passing to the app. The app will take it over, and determine if the TE is overlapping with other or not.

NghiaTranUIT commented on src/context.cc:
```diff
@@ -359,6 +359,7 @@ UIElements UIElements::Reset() {
     render.display_unsynced_items = true;
 
     render.open_time_entry_list = true;
+    render.display_timeline = true;
```
I'd render the timeline by default because the Timeline UI and TimeEntryList are rendering in main app.

NghiaTranUIT commented on src/user.cc:
```diff
@@ -180,15 +182,22 @@ TimeEntry *User::Start(
-        te->SetDurationInSeconds(-now);
-        // dont set Stop, TE is running
-        te->SetStart(now);
+        int seconds = int(ended - started);
```
Introduce the `started` and `ended` to able to create new empty TE at particular time.

NghiaTranUIT commented:
@IndrekV it's ready for reviewing 👍

To specific, I couldn't rebase seamlessly from `master`, since there are a lot of changes and this branch has more than 40+ commits. Thus, I merge directly from `master`, and it significantly reduced the conflicts.

IndrekV commented:
Issues that I found:

- [ ] It's not possible to close edit popup when you open it from timeline view. esc or pressing `x` does not work
- [ ] View does not re-render when I look at it once and then create a time entry. I created a small few second entry and it was not shown in timeline view. I tried switching days but no luck. After app restart, the time entry appeared.
- [ ] For some reason the listing renders the items twice. First set of items do not have duration and when clicking on them the list rerenders correctly

![mac](https://user-images.githubusercontent.com/842229/62216869-3a967400-b3b2-11e9-8215-732bbdb1566a.gif)

NghiaTranUIT commented:
## Changelogs:
- Fix above issues in last review of @IndrekV 
- Fixed missing Duration
- Improve algorithm to calculate and suitable columns for TimelineEntry (In the past, it was responsible by TimelineFlowLayout)
- Find the suitable number of CornerRadius when rendering the TimelineEntry. Ex: Some TimelineEntry is big in 4x Zoom mode, but in 1x zoom, it's too small and the CornerRadius messes up the layout
- Removed old codes.

It's ready to review in second phrase @IndrekV  👍

skel35 commented:
Issues that I found that were not mentioned yet:
- [ ] The "Add time entry" button in gaps should only be shown if there are no time entries at all during this time. For example, there should be no "Add time entry" button here. (The design is not clear about this, but Indrek explained it somewhere).
![image](https://user-images.githubusercontent.com/16451391/62626657-86a46400-b930-11e9-9da1-8917db61abeb.png)
- [ ] Changing the date of an entry in the Timeline tab is very confusing. The re-render works correctly, but the Edit popup remains visible, with an unchanged date. As a simple solution, I suggest hiding the Edit popup after the date was changed.

Also:
- Manual mode doesn't work while on the Timeline tab. This is definitely a separate issue.
Regarding the two above you can fix these here or create new tickets for them.

NghiaTranUIT commented:
@skel35 this PR is ready for third round of review 👍 

## Changelogs
- Fixed above issue 
- The manual issue will take in separate ticket https://github.com/toggl/toggldesktop/issues/3189
